moby/moby
# Return error when Splunk log channel is full in non-blocking mode

**- What I did**

Made the Splunk log driver return an error when the buffer is full in non-blocking mode..

**- How I did it**

The Splunk log driver is now aware of non-blocking mode, in which case it will return an error if the channel is full rather than just block. This will cause the wrapping RingLogger output an error message and increment the log failure metric when dropping messages, instead of them being silently dropped as before. This is similar to how the awslogs driver functions.

This is a very desired feature for us, as we are currently blind to whether or not messages are being dropped when our log sink is not handling the incoming volumes.

I'm very much not a Go developer, and this is my first contribution to Docker, so a constructive and helpful review is appreciated.

**- How to verify it**

This can be verified by running a random container with the Splunk log driver pointing to a Splunk HEC that's slow to respond. Perhaps by putting toxiproxy in between. I simulated it by putting a little sleep before the call to the HEC. Then produce a bunch of logs:

```sh
docker run --log-driver=splunk --log-opt mode=non-blocking --log-opt splunk-token=fake-token --log-opt splunk-url=http://splunk:8088 --log-opt splunk-verify-connection=false  -d -p 80:80 -d nginx
```

Once the channel fills up, you start seeing error logs:

```
ERRO[2021-11-06T15:08:46.472297300Z] Error writing log message                     driver=splunk error="splunk: buffer is full" message=
```

Pulling metrics from the openmetrics endpoint also shows that the `logger_log_write_operations_failed_total` metric gets incremented as expected.

Unfortunately this isn't a _quite_ complete solution, as the Splunk log driver itself has an internal queue that it can drop messages from, which emit logs but don't increment the metrics. Still, this at least gives observability into something that was previously completely silent.

**Description for the changelog:** Splunk Logging Driver: report error log and metrics when dropping logs in non-blocking mode

Obligatory cute animal picture, as requested:

![chomsky](https://user-images.githubusercontent.com/83586/140614745-7331c4b2-6af2-4692-84f4-940316453985.jpeg)
